### ISSUE TITLE
Typescript Error when compiling icon-button

### ISSUE THREAD

**Issue by dawsonc623** [jamesmfriedman/rmwc#587]
- What RMWC Version are you using [major.minor.patch]: 6.0.3

- Name your build system [Webpack, Rollup...]: Webpack

- Describe the bug with as much detail as possible: When trying to compile my project, I get this error:

```
[at-loader] ./node_modules/@rmwc/icon-button/dist/index.d.ts:36:58 
    TS2344: Type '"cite" | "data" | "form" | "label" | "slot" | "span" | "style" | "summary" | "title" | "pattern" | "children" | "accept" | "acceptCharset" | "action" | "allowFullScreen" | "allowTransparency" | ... 342 more ... | "css"' does not satisfy the constraint '"cite" | "data" | "form" | "label" | "slot" | "span" | "style" | "summary" | "title" | "pattern" | "children" | "onChange" | "accept" | "acceptCharset" | "action" | "allowFullScreen" | ... 342 more ... | "onTransitionEndCapture"'.
  Type '"css"' is not assignable to type '"cite" | "data" | "form" | "label" | "slot" | "span" | "style" | "summary" | "title" | "pattern" | "children" | "onChange" | "accept" | "acceptCharset" | "action" | "allowFullScreen" | ... 342 more ... | "onTransitionEndCapture"'.
```
The offending statement is:

```
Pick<React.AllHTMLAttributes<HTMLButtonElement>, "onError" | "className" | "cite" | "data" | "form" | "label" | "span" | "style" | "summary" | "title" | "pattern" | "prefix" | "size" | "children" | "value" | "onBlur" | "onClick" | "onContextMenu" | "onCopy" | "onCut" | "onAuxClick" | "onDoubleClick" | "onDragEnd" | "onDragStart" | "onDrop" | "onFocus" | "onInput" | "onInvalid" | "onKeyDown" | "onKeyPress" | "onKeyUp" | "onMouseDown" | "onMouseUp" | "onPaste" | "onPause" | "onPlay" | "onPointerCancel" | "onPointerDown" | "onPointerUp" | "onRateChange" | "onReset" | "onSeeked" | "onSubmit" | "onTouchCancel" | "onTouchEnd" | "onTouchStart" | "onVolumeChange" | "onAbort" | "onAnimationEnd" | "onAnimationIteration" | "onAnimationStart" | "onCanPlay" | "onCanPlayThrough" | "onDrag" | "onDragEnter" | "onDragExit" | "onDragLeave" | "onDragOver" | "onDurationChange" | "onEmptied" | "onEncrypted" | "onEnded" | "onGotPointerCapture" | "onLoad" | "onLoadedData" | "onLoadedMetadata" | "onLoadStart" | "onLostPointerCapture" | "onMouseMove" | "onMouseOut" | "onMouseOver" | "onPlaying" | "onPointerMove" | "onPointerOut" | "onPointerOver" | "onProgress" | "onScroll" | "onSeeking" | "onStalled" | "onSuspend" | "onTimeUpdate" | "onTouchMove" | "onTransitionEnd" | "onWaiting" | "onWheel" | "onMouseEnter" | "onMouseLeave" | "onPointerEnter" | "onPointerLeave" | "onSelect" | "onBeforeInput" | "onCompositionEnd" | "onCompositionStart" | "onCompositionUpdate" | "id" | "dir" | "slot" | "accept" | "acceptCharset" | "action" | "allowFullScreen" | "allowTransparency" | "alt" | "as" | "async" | "autoComplete" | "autoFocus" | "autoPlay" | "capture" | "cellPadding" | "cellSpacing" | "charSet" | "challenge" | "checked" | "classID" | "cols" | "colSpan" | "content" | "controls" | "coords" | "crossOrigin" | "dateTime" | "default" | "defer" | "disabled" | "download" | "encType" | "formAction" | "formEncType" | "formMethod" | "formNoValidate" | "formTarget" | "frameBorder" | "headers" | "height" | "high" | "href" | "hrefLang" | "htmlFor" | "httpEquiv" | "integrity" | "keyParams" | "keyType" | "kind" | "list" | "loop" | "low" | "manifest" | "marginHeight" | "marginWidth" | "max" | "maxLength" | "media" | "mediaGroup" | "method" | "min" | "minLength" | "multiple" | "muted" | "name" | "nonce" | "noValidate" | "open" | "optimum" | "placeholder" | "playsInline" | "poster" | "preload" | "readOnly" | "rel" | "required" | "reversed" | "rows" | "rowSpan" | "sandbox" | "scope" | "scoped" | "scrolling" | "seamless" | "selected" | "shape" | "sizes" | "src" | "srcDoc" | "srcLang" | "srcSet" | "start" | "step" | "target" | "type" | "useMap" | "width" | "wmode" | "wrap" | "defaultChecked" | "defaultValue" | "suppressContentEditableWarning" | "suppressHydrationWarning" | "accessKey" | "contentEditable" | "contextMenu" | "draggable" | "hidden" | "lang" | "spellCheck" | "tabIndex" | "translate" | "radioGroup" | "role" | "about" | "datatype" | "inlist" | "property" | "resource" | "typeof" | "vocab" | "autoCapitalize" | "autoCorrect" | "autoSave" | "color" | "itemProp" | "itemScope" | "itemType" | "itemID" | "itemRef" | "results" | "security" | "unselectable" | "inputMode" | "is" | "aria-activedescendant" | "aria-atomic" | "aria-autocomplete" | "aria-busy" | "aria-checked" | "aria-colcount" | "aria-colindex" | "aria-colspan" | "aria-controls" | "aria-current" | "aria-describedby" | "aria-details" | "aria-disabled" | "aria-dropeffect" | "aria-errormessage" | "aria-expanded" | "aria-flowto" | "aria-grabbed" | "aria-haspopup" | "aria-hidden" | "aria-invalid" | "aria-keyshortcuts" | "aria-label" | "aria-labelledby" | "aria-level" | "aria-live" | "aria-modal" | "aria-multiline" | "aria-multiselectable" | "aria-orientation" | "aria-owns" | "aria-placeholder" | "aria-posinset" | "aria-pressed" | "aria-readonly" | "aria-relevant" | "aria-required" | "aria-roledescription" | "aria-rowcount" | "aria-rowindex" | "aria-rowspan" | "aria-selected" | "aria-setsize" | "aria-sort" | "aria-valuemax" | "aria-valuemin" | "aria-valuenow" | "aria-valuetext" | "dangerouslySetInnerHTML" | "onCopyCapture" | "onCutCapture" | "onPasteCapture" | "onCompositionEndCapture" | "onCompositionStartCapture" | "onCompositionUpdateCapture" | "onFocusCapture" | "onBlurCapture" | "onChangeCapture" | "onBeforeInputCapture" | "onInputCapture" | "onResetCapture" | "onSubmitCapture" | "onInvalidCapture" | "onLoadCapture" | "onErrorCapture" | "onKeyDownCapture" | "onKeyPressCapture" | "onKeyUpCapture" | "onAbortCapture" | "onCanPlayCapture" | "onCanPlayThroughCapture" | "onDurationChangeCapture" | "onEmptiedCapture" | "onEncryptedCapture" | "onEndedCapture" | "onLoadedDataCapture" | "onLoadedMetadataCapture" | "onLoadStartCapture" | "onPauseCapture" | "onPlayCapture" | "onPlayingCapture" | "onProgressCapture" | "onRateChangeCapture" | "onSeekedCapture" | "onSeekingCapture" | "onStalledCapture" | "onSuspendCapture" | "onTimeUpdateCapture" | "onVolumeChangeCapture" | "onWaitingCapture" | "onAuxClickCapture" | "onClickCapture" | "onContextMenuCapture" | "onDoubleClickCapture" | "onDragCapture" | "onDragEndCapture" | "onDragEnterCapture" | "onDragExitCapture" | "onDragLeaveCapture" | "onDragOverCapture" | "onDragStartCapture" | "onDropCapture" | "onMouseDownCapture" | "onMouseMoveCapture" | "onMouseOutCapture" | "onMouseOverCapture" | "onMouseUpCapture" | "onSelectCapture" | "onTouchCancelCapture" | "onTouchEndCapture" | "onTouchMoveCapture" | "onTouchStartCapture" | "onPointerDownCapture" | "onPointerMoveCapture" | "onPointerUpCapture" | "onPointerCancelCapture" | "onPointerEnterCapture" | "onPointerLeaveCapture" | "onPointerOverCapture" | "onPointerOutCapture" | "onGotPointerCaptureCapture" | "onLostPointerCaptureCapture" | "onScrollCapture" | "onWheelCapture" | "onAnimationStartCapture" | "onAnimationEndCapture" | "onAnimationIterationCapture" | "onTransitionEndCapture" | "css">
```
If I remove the ` | "css"` off of the end, the code compiles.

- What happened, and what was supposed to happen: The compilation failed when it should have succeeded without need to alter the library code.

**Comment by jamesmfriedman** [comment]
Have you tried upgrading your version of @types/react to the latest? Also what version of typescript are you using?

**Comment by dawsonc623** [comment]
This is a fresh project I just spun up this morning. Dependencies:

```
  "dependencies": {
    "@rmwc/top-app-bar": "^6.0.3",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "react-router-dom": "^5.1.2"
  },
  "devDependencies": {
    "@types/react": "^16.9.27",
    "@types/react-dom": "^16.9.5",
    "@types/react-router-dom": "^5.1.3",
    "@typescript-eslint/eslint-plugin": "^2.25.0",
    "@typescript-eslint/parser": "^2.25.0",
    "awesome-typescript-loader": "^5.2.1",
    "css-loader": "^3.4.2",
    "eslint": "^6.8.0",
    "eslint-plugin-import": "^2.20.2",
    "eslint-plugin-react": "^7.19.0",
    "eslint-plugin-react-hooks": "^3.0.0",
    "html-webpack-plugin": "^4.0.3",
    "node-sass": "^4.13.1",
    "sass-loader": "^8.0.2",
    "style-loader": "^1.1.3",
    "typescript": "^3.8.3",
    "webpack": "^4.42.1",
    "webpack-cli": "^3.3.11",
    "webpack-dev-server": "^3.10.3",
    "webpack-merge": "^4.2.2"
  }
```

**Comment by jamesmfriedman** [comment]
Thank you, that’s very helpful. I’ll take a look in a bit. Doesn’t make any sense to me why it wouldn’t compile.

Definition files generated by typescript have a habit of flattening types that use any helpers like Omit instead of using the original type declaration. It may be an issue due to that, although the props interface for V6 remains largely unchanged.

**Comment by dawsonc623** [comment]
It did not make a lot of sense to me either, but I am far from an expert in this area. I can try to keep digging on my end, too. The code is dirt simple right now. The only component is the main app component with the `TopAppBar` from this project and a basic React Router set up just rendering "Hello World".

If you have anything else you want me to check, let me know.

**Comment by jamesmfriedman** [comment]
I just fired up a new create-react-app project with the top app bar and don't have any issue building or running the dev server. If you want to share your example repo I could help dig in there.

**Comment by dawsonc623** [comment]
https://github.com/dawsonc623/digimon-world-2003

Sorry for the delay. I did not have the project in Git yet.

**Comment by jamesmfriedman** [comment]
I checked out the repro. I'm really at a loss here. I understand what Typescript is doing here, but I definitely don't see why this causes an error. Is it possible something with the configuration?

**Comment by jamesmfriedman** [comment]
[foo.zip](https://github.com/jamesmfriedman/rmwc/files/4400809/foo.zip)

Here is a shell of a project with create-react-app, with the app bar installed. Compiles fine, type checking also working fine.

**Comment by jamesmfriedman** [comment]
I audited the generated type definitions and found a way to cast certain offending components with the proper types so they don't flatten their types in the emitted declaration files (causing the long chain you were seeing). 

You have other compilation issues in the sample you sent me, but RMWC no longer appears to be the culprit. Fixed in 6.0.5

**Comment by dawsonc623** [comment]
Thank you. I have confirmed it works for my project now, too.

**Comment by jamesmfriedman** [comment]
Great. Thanks for the confirm! Have fun with the Digimon app ;)

**Comment by dawsonc623** [comment]
@jamesmfriedman This issue popped back up in two places:

* `@rmwc/base/dist/component.d.ts:15:4`
* `@rmwc/dialog/dist/dialog.d.ts:64:173`

I can create a branch on my project's repo so you can pull it down and see it if you cannot reproduce it yourself.

**Comment by dawsonc623** [comment]
https://github.com/elyctech/sword-of/tree/problematic-code

That is a Lerna/Yarn project, but neither package has been deployed yet. The easiest way to get the project up and running is to run `yarn link` in `packages/sword-of-engine` and `yarn link sword-of-engine` in `packages/sword-of-editor`.

**Comment by jamesmfriedman** [comment]
Cool, I’ll reopen. I went and audited all of the external definitions prior. Interestingly enough I still don’t experience the problems with create react app so it is something that is build system dependent. Either way it is helpful to improve the exported definitions.

**Comment by jamesmfriedman** [comment]
Fixed in 6.0.12

**Comment by dawsonc623** [comment]
Yeah, not sure about create react app. I have a project scaffold I use instead. It is based on the create react app structure and configuration, but I changed some things to fit my liking. I am not exactly sure what about my changes bring up these problems, though.
